crates/task-impls/src/transactions.rs

```diff
@@ -312,93 +327,174 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, Ver: StaticVersionType>
-                (u128::from(l.offered_fee) * u128::from(r.block_size))
-                    .cmp(&(u128::from(r.offered_fee) * u128::from(l.block_size)))
-            })
-            .context("no available blocks")?;
-        tracing::debug!("Selected block: {block_info:?}");
-
-        // Verify signature over chosen block.
-        if !block_info.sender.validate_block_info_signature(
-            &block_info.signature,
-            block_info.block_size,
-            block_info.offered_fee,
-            &block_info.block_hash,
-        ) {
-            bail!("Failed to verify available block info response message signature");
-        }
+        // Create a future for every builder
+        let mut available_blocks = {
+            let tasks = self
+                .builder_clients
+                .iter()
+                .enumerate()
+                .map(|(idx, client)| async move {
+                    client
+                        .available_blocks(
+                            parent_comm,
+                            view_number.u64(),
+                            self.public_key.clone(),
+                            parent_comm_sig,
+                        )
+                        .await
+                        .map(move |blocks| {
+                            blocks.into_iter().map(move |block_info| (block_info, idx))
+                        })
+                })
+                .collect::<FuturesUnordered<_>>();
+            let mut results = Vec::with_capacity(self.builder_clients.len());
+
+            // Instant we start querying builders for available blocks
+            let query_start = Instant::now();
+
+            // First we complete the query to the fastest fraction of the builders
+            let threshold = (self.builder_clients.len() * BUILDER_MAIN_BATCH_THRESHOLD_DIVIDEND)
+                .div_ceil(BUILDER_MAIN_BATCH_THRESHOLD_DIVISOR);
+            let mut tasks = tasks.take(threshold);
+            while let Some(result) = tasks.next().await {
+                results.push(result);
+                if query_start.elapsed() > BUILDER_MAIN_BATCH_CUTOFF {
+                    break;
+                }
+            }
 
-        let request_signature = <<TYPES as NodeType>::SignatureKey as SignatureKey>::sign(
-            &self.private_key,
-            block_info.block_hash.as_ref(),
-        )
-        .context("signing block hash")?;
+            // Then we query the rest, alotting additional `elapsed * BUILDER_ADDITIONAL_TIME_MULTIPLIER`
+            // for them to respond. There's a fixed floor of `BUILDER_MINIMUM_QUERY_TIME` for both
+            // phases
+            let timeout = async_sleep(std::cmp::max(
+                query_start
+                    .elapsed()
+                    .mul_f32(BUILDER_ADDITIONAL_TIME_MULTIPLIER),
+                BUILDER_MINIMUM_QUERY_TIME - query_start.elapsed(),
+            ));
+            futures::pin_mut!(timeout);
```

Comment:
The `pin_mut` here is not immediately clear at first glance. Can we get a comment explaining why we're using it here?